server/player/player.go

```diff
@@ -1044,6 +1044,37 @@ func (p *Player) StopSwimming() {
 	p.updateState()
 }
 
+// StartCrawling makes the player start crawling if it is not currently doing so. If the player is sneaking
+// while StartCrawling is called, the sneaking is stopped.
+func (p *Player) StartCrawling() {
+	var isAir bool
+	for _, corner := range p.Type().BBox(p).Translate(p.Position()).Corners() {
+		_, isAir = p.World().Block(cube.PosFromVec3(corner).Add(cube.Pos{0, 1, 0})).(block.Air)
+		if !isAir {
+			break
+		}
+	}
+
+	if !isAir && !p.crawling.CompareAndSwap(false, true) {
+		return
+	}
```

Comment:
This can be slightly simplified to
```go
for _, corner := range p.Type().BBox(p).Translate(p.Position()).Corners() {
	_, isAir = p.World().Block(cube.PosFromVec3(corner).Add(cube.Pos{0, 1, 0})).(block.Air)
	if !isAir {
		if !p.crawling.CompareAndSwap(false, true) {
			return
		}
		break
	}
}
```